README.md
# arXiv

[![License](https://img.shields.io/badge/License-MIT%20%26%20Apache%202.0-blue)](#license)
[![docs.rs](https://img.shields.io/docsrs/arxiv/latest)](https://docs.rs/arxiv/)
[![CI](https://github.com/acmuta-research/arxiv-rs/actions/workflows/main.yml/badge.svg)](https://github.com/acmuta-research/arxiv-rs/actions/workflows/main.yml)
[![Security audit](https://github.com/acmuta-research/arxiv-rs/actions/workflows/security-audit.yml/badge.svg)](https://github.com/acmuta-research/arxiv-rs/actions/workflows/security-audit.yml)
[![codecov](https://codecov.io/gh/acmuta-research/arxiv-rs/branch/main/graph/badge.svg?token=6ZSIWAQTHU)](https://codecov.io/gh/acmuta-research/arxiv-rs)

A Rust library for parsing `arXiv` identifiers and references.

## Install

Run the following command in the terminal:

```shell
cargo add arxiv
```

Or, add this to `Cargo.toml`:

```shell
[dependencies]
arxiv = "0.1"
```

## Usage

```rust
use std::str::FromStr;
use arxiv::{ArxivId, ArxivStamp};

// Parse an arXiv identifier
let id = ArxivId::from_str("arXiv:9912.12345v2").unwrap();
assert_eq!(id.month, 12);
assert_eq!(id.year, 2099);
assert_eq!(id.number, "12345");
assert_eq!(id.version, Some(2));

// Parse an arXiv stamp
let stamp = ArxivStamp::from_str("arXiv:0706.0001v1 [q-bio.CB] 1 Jun 2007").unwrap();
assert_eq!(stamp.category, "q-bio.CB");
assert_eq!(stamp.submitted.year(), 2007);
```

## License

Licensed under either of

* Apache License, Version 2.0 ([`LICENSE-APACHE`](LICENSE-APACHE) or <http://www.apache.org/licenses/LICENSE-2.0>)
* MIT license ([`LICENSE-MIT`](LICENSE-MIT) or <http://opensource.org/licenses/MIT>)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Changelog

## 0.1.0 (2023-03-28)

- Initial release of library


src/stamp.rs
use crate::{ArxivId, ArxivIdError};
use std::error::Error;
use std::fmt::{Display, Formatter, Result as FmtResult};
use std::str::FromStr;
use time::error::Parse as TimeParseError;
use time::macros::format_description;
use time::{Date, Month};

/// Convenient type alias for a [`Result`] holding either an [`ArxivStamp`] or [`ArxivStampError`]
pub type ArxivStampResult = Result<ArxivStamp, ArxivStampError>;
type CategoryResult = Result<String, ArxivStampError>;
type DateParseResult = Result<Date, TimeParseError>;

/// An error that can occur when parsing and validating arXiv stamps
///
/// # Examples
/// ```
/// use std::str::FromStr;
/// use arxiv::ArxivStamp;
///
/// let stamp = ArxivStamp::from_str("arXiv:2001.00001 [cs.LG] 1 Jan 2000");
/// ```
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ArxivStampError {
	InvalidArxivId(ArxivIdError),
	InvalidDate(TimeParseError),
	InvalidCategory,
	NotEnoughComponents,
}

impl Error for ArxivStampError {}

impl Display for ArxivStampError {
	fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
		match self {
			Self::InvalidArxivId(e) => write!(f, "Invalid arXiv ID: {}", e),
			Self::InvalidDate(e) => write!(f, "Invalid date: {}", e),
			Self::InvalidCategory => write!(f, "Invalid category"),
			Self::NotEnoughComponents => write!(f, "Not enough components"),
		}
	}
}

/// A stamp that is added onto the side of PDF version of arXiv articles
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ArxivStamp {
	pub id: ArxivId,
	pub category: String,
	pub submitted: Date,
}

impl ArxivStamp {
	pub(crate) const TOKEN_SPACE: char = ' ';

	/// Manually create a new [`ArxivStamp`] from the given components. This is useful
	/// if you want to dynamically create a stamp from user input; otherwise, it is
	/// recommended to use [`ArxivStamp::from_str()`] instead since it can be quite verbose.
	///
	/// # Examples
	/// ```
	/// use time::{Date, Month};
	/// use arxiv::{ArxivId, ArxivStamp};
	///
	/// let stamp = ArxivStamp::new(
	///    ArxivId::try_latest(2011, 1, String::from("00001")).unwrap(),
	///    String::from("cs.LG"),
	///    Date::from_calendar_date(2011, Month::January, 1).unwrap()
	/// );
	/// ```
	#[inline]
	pub const fn new(id: ArxivId, category: String, submitted: Date) -> Self {
		Self {
			id,
			category,
			submitted,
		}
	}
}

impl Display for ArxivStamp {
	fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
		write!(
			f,
			"{} [{}] {} {} {}",
			self.id,
			self.category,
			self.submitted.day(),
			month_as_abbr(self.submitted.month()),
			self.submitted.year()
		)
	}
}

impl FromStr for ArxivStamp {
	type Err = ArxivStampError;

	fn from_str(s: &str) -> Result<Self, Self::Err> {
		let parts = s.splitn(2, ArxivStamp::TOKEN_SPACE).collect::<Vec<&str>>();

		if parts.len() == 1 {
			return Err(ArxivStampError::NotEnoughComponents);
		}

		let arxiv_id = ArxivId::from_str(parts[0]);
		if let Err(e) = arxiv_id {
			return Err(ArxivStampError::InvalidArxivId(e));
		}

		// category is opitional, so we need to check if the second part is a category
		// and decide which index to use to parse each component
		let part2_is_category = parts[1].starts_with('[');
		let mut category: CategoryResult = Ok(String::new());
		let date: DateParseResult;

		if part2_is_category {
			let category_date = parts[1]
				.splitn(2, ArxivStamp::TOKEN_SPACE)
				.collect::<Vec<&str>>();

			category = parse_category(category_date[0]);
			date = parse_date(category_date[1]);
		} else {
			date = parse_date(parts[1]);
		}

		// validate the category and date components
		// converting this to category? will cause the value to move,
		#[allow(clippy::question_mark)]
		if let Err(e) = category {
			return Err(e);
		}

		if let Err(e) = date {
			return Err(ArxivStampError::InvalidDate(e));
		}

		// if we got this far, we can safely unwrap the results
		Ok(Self::new(arxiv_id.unwrap(), category.unwrap(), date.unwrap()))
	}
}

fn parse_category(s: &str) -> CategoryResult {
	if brackets_match(s) {
		Ok(s[1..s.len() - 1].to_string())
	} else {
		Err(ArxivStampError::InvalidCategory)
	}
}

fn brackets_match(s: &str) -> bool {
	s.starts_with('[') && s.ends_with(']')
		|| s.starts_with('(') && s.ends_with(')')
		|| s.starts_with('{') && s.ends_with('}')
}

const fn month_as_abbr<'a>(month: Month) -> &'a str {
	match month {
		Month::January => "Jan",
		Month::February => "Feb",
		Month::March => "Mar",
		Month::April => "Apr",
		Month::May => "May",
		Month::June => "June",
		Month::July => "July",
		Month::August => "Aug",
		Month::September => "Sept",
		Month::October => "Oct",
		Month::November => "Nov",
		Month::December => "Dec",
	}
}

/// Parses a date in the form of "1 Jan 2000", where:
///  - the day is a number without zero padding
///  - the month is the first three letters of the full month name
///  - the year is a 4-digit number
///
/// See also: [`time` documentation for format descriptions][time-format-desc]
///
/// [time-format-desc]: https://time-rs.github.io/book/api/format-description.html
fn parse_date(date_str: &str) -> DateParseResult {
	Date::parse(date_str, &format_description!("[day padding:none] [month repr:short] [year]"))
}

#[cfg(test)]
mod tests {
	use super::*;
	use std::str::FromStr;
	use time::error::ParseFromDescription;
	use time::Date;

	#[test]
	fn display_stamp() {
		let stamp = ArxivStamp::new(
			ArxivId::from_str("arXiv:2011.00001").unwrap(),
			String::from("cs.LG"),
			Date::from_calendar_date(2011, Month::January, 1).unwrap(),
		);
		assert_eq!(stamp.to_string(), "arXiv:2011.00001 [cs.LG] 1 Jan 2011");
	}

	#[test]
	fn parse_stamp() {
		let stamp = "arXiv:2001.00001 [cs.LG] 1 Jan 2000";
		let parsed = ArxivStamp::from_str(stamp);
		assert_eq!(
			parsed,
			Ok(ArxivStamp::new(
				ArxivId::from_str("arXiv:2001.00001").unwrap(),
				String::from("cs.LG"),
				Date::from_calendar_date(2000, Month::January, 1).unwrap()
			))
		);
	}

	#[test]
	fn parse_stamp_without_category() {
		let stamp = "arXiv:2001.00001 1 Jan 2000";
		let parsed = ArxivStamp::from_str(stamp);
		assert_eq!(
			parsed,
			Ok(ArxivStamp::new(
				ArxivId::from_str("arXiv:2001.00001").unwrap(),
				String::new(),
				Date::from_calendar_date(2000, Month::January, 1).unwrap()
			))
		);
	}

	#[test]
	fn parse_stamp_empty() {
		let stamp = "";
		let parsed = ArxivStamp::from_str(stamp);
		assert_eq!(parsed, Err(ArxivStampError::NotEnoughComponents));
	}

	#[test]
	fn parse_stamp_not_enough_components() {
		let stamp = "arXiv:2001.00001";
		let parsed = ArxivStamp::from_str(stamp);
		assert_eq!(parsed, Err(ArxivStampError::NotEnoughComponents));
	}

	#[test]
	fn parse_stamp_invalid_category() {
		let stamp = "arXiv:2001.00001 [cs.LG 1 Jan 2000";
		let parsed = ArxivStamp::from_str(stamp);
		assert_eq!(parsed, Err(ArxivStampError::InvalidCategory));
	}

	#[test]
	fn parse_stamp_invalid_date_day() {
		let stamp = "arXiv:2001.00001 [cs.LG] 32 Jan 2000";
		let parsed = ArxivStamp::from_str(stamp);

		// hack to get a ComponentRange error
		let date = parse_date("32 Jan 2000").unwrap_err();

		assert_eq!(parsed, Err(ArxivStampError::InvalidDate(date)));
	}

	#[test]
	fn parse_stamp_invalid_date_month() {
		let stamp = "arXiv:2001.00001 [cs.LG] 1 Zan 2000";
		let parsed = ArxivStamp::from_str(stamp);
		assert_eq!(parsed, Err(invalid_date_component("month")));
	}

	#[test]
	fn parse_stamp_invalid_date_year() {
		let stamp = "arXiv:2001.00001 [cs.LG] 1 Jan 200";
		let parsed = ArxivStamp::from_str(stamp);
		assert_eq!(parsed, Err(invalid_date_component("year")));
	}

	fn invalid_date_component(component: &'static str) -> ArxivStampError {
		ArxivStampError::InvalidDate(TimeParseError::ParseFromDescription(
			ParseFromDescription::InvalidComponent(component),
		))
	}
}


src/identifier.rs
use std::error::Error;
use std::fmt::{Display, Formatter, Result as FmtResult};
use std::str::FromStr;

/// Convenient type alias for a [`Result`] holding either an [`ArxivId`] or [`ArxivIdError`]
pub type ArxivIdResult = Result<ArxivId, ArxivIdError>;

/// An error that can occur when parsing and validating arXiv identifiers
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ArxivIdError {
	/// A generic parsing syntax error
	Syntax,
	/// An ivalid month outside of the inclusive [1, 12] interval
	InvalidMonth,
	/// An invalid year outside of the inclusive [2007, 2099] interval
	InvalidYear,
	/// An invalid year outside of the inclusive [1, 99999] interval
	InvalidId,
}

impl Error for ArxivIdError {}

impl Display for ArxivIdError {
	fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
		match self {
			Self::Syntax => write!(f, "There was a syntax error; an ArXiv identifier must conform to the schema of arXiv:YYMM.number{{vV}}."),
			Self::InvalidMonth => write!(f, "A valid month must be between 1 and 12."),
			Self::InvalidYear => write!(f, "A valid year must be be between 2007 and 2099."),
			Self::InvalidId => write!(f, "A valid identifier must be between 1 and 99999."),
		}
	}
}

/// A unique identifier for articles published on arXiv.org
///
/// See also: [Official arXiv.org documentation][arxiv-docs]
///
/// # Examples
/// ```
/// use std::str::FromStr;
/// use arxiv::ArxivId;
///
/// let id = ArxivId::from_str("arXiv:2001.00001");
/// ```
///
/// [arxiv-docs]: https://info.arxiv.org/help/arxiv_identifier.html
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ArxivId {
	pub year: u16,
	pub month: u8,
	pub number: String,
	pub version: Option<u8>,
}

impl ArxivId {
	pub const MIN_YEAR: u16 = 2007u16;
	pub const MAX_YEAR: u16 = 2099u16;
	pub const MIN_MONTH: u8 = 1u8;
	pub const MAX_MONTH: u8 = 12u8;
	pub(crate) const TOKEN_COLON: char = ':';
	pub(crate) const TOKEN_DOT: char = '.';
	pub(crate) const TOKEN_VERSION: char = 'v';

	/// This allows manually creating an [`ArxivId`] from the given components without any
	/// validation. Only do this if you have already verified that the components are valid.
	///
	/// # Examples
	/// ```
	/// use arxiv::ArxivId;
	///
	/// let id = ArxivId::new_raw(2011, 1, String::from("00001"), Some(1));
	/// ```
	#[inline]
	pub const fn new_raw(year: u16, month: u8, number: String, version: Option<u8>) -> Self {
		Self {
			year,
			month,
			number,
			version,
		}
	}

	/// This allows manually creating an [`ArxivId`] from the given components without any version
	/// (assuming it is the latest version). Only do this if you have already verified that the
	/// components are valid.
	///
	/// # Examples
	/// ```
	/// use arxiv::ArxivId;
	///
	/// let id = ArxivId::new_latest(2011, 1, String::from("00001"));
	/// ```
	#[inline]
	pub const fn new_latest(year: u16, month: u8, id: String) -> Self {
		Self::new_raw(year, month, id, None)
	}

	/// This allows manually creating an [`ArxivId`] from the given components with a version, and
	/// will also validate each component for correctness. If any component is invalid, it will return
	/// an [`ArxivIdError`].
	///
	/// # Examples
	/// ```
	/// use arxiv::{ArxivId, ArxivIdError};
	///
	/// let id = ArxivId::try_new(2011, 1, String::from("00001"), Some(1));
	/// ```
	pub fn try_new(year: u16, month: u8, number: String, version: Option<u8>) -> ArxivIdResult {
		if !(Self::MIN_YEAR..=Self::MAX_YEAR).contains(&year) {
			return Err(ArxivIdError::InvalidYear);
		}

		if !(1..=12).contains(&month) {
			return Err(ArxivIdError::InvalidMonth);
		}

		if number.len() < 4 || number.len() > 5 {
			return Err(ArxivIdError::InvalidId);
		}

		Ok(Self::new_raw(year, month, number, version))
	}

	/// This allows manually creating an [`ArxivId`] from the given components without a version
	/// (assuming it is the latest version), and will also validate each component for correctness.
	/// If any component is invalid, it will return an [`ArxivIdError`].
	///
	/// # Examples
	/// ```
	/// use arxiv::{ArxivId, ArxivIdError};
	///
	/// let id = ArxivId::try_latest(2011, 1, String::from("00001"));
	/// ```
	#[inline]
	pub fn try_latest(year: u16, month: u8, number: String) -> ArxivIdResult {
		Self::try_new(year, month, number, None)
	}

	/// Whether or not the identifier refers to the most recent version of the arXiv article
	#[inline]
	pub const fn is_latest(&self) -> bool {
		self.version.is_none()
	}

	/// Sets the version of the arXiv article.
	///
	/// # Examples
	/// ```
	/// use std::str::FromStr;
	/// use arxiv::ArxivId;
	///
	/// let mut id = ArxivId::from_str("arXiv:2001.00001").unwrap();
	/// id.set_version(1);
	/// ```
	#[inline]
	pub fn set_version(&mut self, version: u8) {
		self.version = Some(version);
	}

	/// Sets the version of the arXiv article to the latest version.
	///
	/// # Examples
	/// ```
	/// use std::str::FromStr;
	/// use arxiv::ArxivId;
	///
	/// let mut id = ArxivId::from_str("arXiv:2001.00001v1").unwrap();
	/// id.set_latest();
	/// ```
	#[inline]
	pub fn set_latest(&mut self) {
		self.version = None;
	}
}

impl Display for ArxivId {
	fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
		let mut binding = self.year.to_string();
		let (_, half_year) = binding.as_mut_str().split_at(2);

		if self.number.len() == 4usize {
			write!(f, "arXiv:{:02}{:02}.{:04}", half_year, self.month, self.number)
		} else {
			write!(f, "arXiv:{:02}{:02}.{:05}", half_year, self.month, self.number)
		}
	}
}

impl FromStr for ArxivId {
	type Err = ArxivIdError;

	fn from_str(value: &str) -> Result<Self, Self::Err> {
		// break down the arxiv string into its components
		let parts: Vec<&str> = value.split(ArxivId::TOKEN_COLON).collect();
		if parts.len() != 2 || parts[0] != "arXiv" {
			return Err(ArxivIdError::Syntax);
		}

		let inner_parts: Vec<&str> = parts[1].split(ArxivId::TOKEN_DOT).collect();
		if inner_parts.len() != 2 {
			return Err(ArxivIdError::Syntax);
		}

		// validate and compose the final Arxiv struct
		let year = inner_parts[0][0..2].parse::<u16>();
		let month = inner_parts[0][2..4].parse::<u8>();
		if year.is_err() || month.is_err() {
			return Err(ArxivIdError::Syntax);
		}

		let (number, version) = parse_numbervv(inner_parts[1]);

		ArxivId::try_new(year.unwrap() + 2000, month.unwrap(), number, version)
	}
}

/// Parses a string in the format of "number{vV}",
/// where:
/// - `number` is a unique integer up 4 to 5 digits
/// - `{vV}` (optional): a `v` literal followed by 1 or more digits
fn parse_numbervv(s: &str) -> (String, Option<u8>) {
	let parts = s.split(ArxivId::TOKEN_VERSION).collect::<Vec<&str>>();

	let number = String::from(parts[0]);
	let mut version: Option<u8> = None;

	if parts.len() == 1 {
		return (number, version);
	}

	let parsed_version = parts[1].parse::<u8>();
	if let Ok(t) = parsed_version {
		version = Some(t);
	}

	(number, version)
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn parse_arxiv_empty() {
		assert_eq!(ArxivId::from_str(""), Err(ArxivIdError::Syntax));
	}

	#[test]
	fn parse_arxiv_without_version() {
		assert_eq!(
			ArxivId::from_str("arXiv:1501.00001"),
			Ok(ArxivId::new_latest(2015, 1, String::from("00001")))
		);
	}

	#[test]
	fn parse_arxiv_with_version() {
		assert_eq!(
			ArxivId::from_str("arXiv:9912.12345v2"),
			Ok(ArxivId::new_raw(2099, 12, String::from("12345"), Some(2)))
		)
	}

	#[test]
	fn arxiv_as_string_number4digits() {
		assert_eq!(
			ArxivId::new_latest(2014, 1, String::from("7878")).to_string(),
			String::from("arXiv:1401.7878")
		);
		assert_eq!(
			ArxivId::new_latest(2014, 12, String::from("7878")).to_string(),
			String::from("arXiv:1412.7878")
		);
	}

	#[test]
	fn arxiv_as_string_number5digits() {
		assert_eq!(
			ArxivId::new_latest(2014, 1, String::from("00008")).to_string(),
			String::from("arXiv:1401.00008")
		);
		assert_eq!(
			ArxivId::new_latest(2014, 12, String::from("00008")).to_string(),
			String::from("arXiv:1412.00008")
		);
	}

	#[test]
	fn parse_arxiv_invalid_year() {
		assert_eq!(
			ArxivId::try_latest(2006, 1, String::from("00001")),
			Err(ArxivIdError::InvalidYear)
		);
	}

	#[test]
	fn parse_arxiv_invalid_month() {
		assert_eq!(
			ArxivId::try_latest(2007, u8::MAX, String::from("00001")),
			Err(ArxivIdError::InvalidMonth)
		)
	}

	#[test]
	fn parse_arxiv_invalid_id() {
		assert_eq!(ArxivId::try_latest(2007, 11, String::new()), Err(ArxivIdError::InvalidId))
	}
}
